pyvrp/cpp/repair/greedy_repair.cpp

```diff
@@ -0,0 +1,164 @@
+#include "greedy_repair.h"
+
+#include "TimeWindowSegment.h"
+#include "search/Route.h"
+
+#include <cassert>
+
+using pyvrp::CostEvaluator;
+using pyvrp::DynamicBitset;
+using pyvrp::ProblemData;
+using pyvrp::Solution;
+using pyvrp::search::n;
+using pyvrp::search::p;
+using pyvrp::search::Route;
+
+using Clients = std::vector<Route::Node>;
+using Routes = std::vector<Route>;
+using SolRoutes = std::vector<Solution::Route>;
+using TWS = pyvrp::TimeWindowSegment;
+
+namespace
+{
+// Populate the given clients and routes vectors with routes from the solution.
+void setupRoutes(Clients &clients,
+                 Routes &routes,
+                 SolRoutes const &solRoutes,
+                 ProblemData const &data)
+{
+    assert(clients.empty() && routes.empty());
+
+    // Doing this avoids re-allocations, which would break the pointer structure
+    // that Route and Route::Node use.
+    clients.reserve(data.numClients() + 1);
+    routes.reserve(solRoutes.size());
```

Comment:
Should this be `routes.reserve(data.numVehicles)` instead be consistent with LS?

Reply:
Not to be consistent with LS per se, but allowing insertions into empty routes doesn't sound like a bad idea if it results in the lowest possible delta cost.

Reply:
I think this is a pretty bad idea, since our first use case is in SREX where we probably don't want to increase the number of routes if we can avoid it. I've bound the alternative overload with routes as well, to add the empty route test. That makes it possible for a user to do what you propose here as well, by providing a list of routes that additionally contain the empty routes. But having the option to restrict the route set can be useful to avoid increasing the number of routes I think.

(if not it'd also defeat the purpose of how we handle empty route moves in the LS, because they'd  be done by SREX anyway)

Reply:
OK, let's keep it like this then!